cpp/src/structs/utilities.cu

```diff
@@ -289,6 +294,140 @@ std::unique_ptr<column> superimpose_nulls(bitmask_type const* null_mask,
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate new device memory
+ * @return A new column with potentially new null mask
+ */
+std::vector<std::unique_ptr<column>> superimpose_nulls(std::vector<bitmask_type const*> null_masks,
+                                                       std::vector<std::unique_ptr<column>> inputs,
+                                                       rmm::cuda_stream_view stream,
+                                                       rmm::device_async_resource_ref mr)
+{
+  CUDF_FUNC_RANGE();
+
+  CUDF_EXPECTS(null_masks.size() == inputs.size(),
+               "The number of null masks to apply must match the number of input columns");
+
+  auto const num_rows = inputs[0]->size();
+  std::vector<bitmask_type const*> sources;
+  std::vector<size_type> segment_offsets;
+  std::vector<bitmask_type const*> path;
+
+  // This recursive function navigates the column hierarchy and for each path in the tree, it
+  // collects all null masks that need to be combined for each column in the hierarchy
+  std::function<void(column & input)> populate_segmented_sources =
+    [&populate_segmented_sources, &path, &sources, &segment_offsets](column& input) -> void {
+    if (input.type().id() != cudf::type_id::EMPTY) {
+      // EMPTY columns should not have a null mask,
+      // so don't superimpose null mask on empty columns.
+      if (input.nullable()) {
+        // Add this column's null mask to the current path
+        path.push_back(input.mutable_view().null_mask());
+      }
+
+      // Add all null masks in current path to sources
+      sources.insert(sources.end(), path.begin(), path.end());
+      segment_offsets.push_back(path.size());
+
+      // For struct columns, recursively process all children
+      if (input.type().id() == cudf::type_id::STRUCT) {
+        for (int i = 0; i < input.num_children(); i++) {
+          populate_segmented_sources(input.child(i));
+        }
+      }
+
+      // Backtrack: remove this column's mask from path when done
+      if (input.nullable()) path.pop_back();
+    }
+  };
+
+  // Track/mark the starting position in segment_offsets for each top-level column
+  std::vector<size_type> column_markers;
+  column_markers.push_back(0);
+
+  // Process each top-level column in the inputs vector
+  for (size_t c = 0; c < null_masks.size(); c++) {
+    // Start with the external null mask for this column
+    path.push_back(null_masks[c]);
+
+    // Collect all null masks for this column and its descendants
+    populate_segmented_sources(*(inputs[c]));
+    path.pop_back();
+
+    // Record where this column's segments end in the segment_offsets array
+    column_markers.push_back(segment_offsets.size());
+  }
+
+  // Convert segment_offsets from segment sizes to cumulative offsets
+  {
+    auto total_sum = std::accumulate(segment_offsets.begin(), segment_offsets.end(), 0);
```

Comment:
Suggesting a neat (completely optional) trick here. If you push a 0 to `segment_offsets` when constructed above, you can just do an `std::inclusive_scan(begin+1, end)` here to get the desired result. total_sum will just be `segment_offsets.back()` if needed